dilation-protocol.md

```diff
@@ -1,371 +1,280 @@
-derived key)
-
-(TODO: reduce the number of round-trip stalls here, I've added too many)
-
-Each side is in the "connecting" state (which encompasses both making
-connection attempts and having an established connection) starting with the
-receipt of a `please` message and a local `w.dilate()` call. The
-Leader remains in that state until it abandons the connection and sends a
-`reconnect` message, at which point it remains in the "flushing" state until
-the Follower's `reconnecting` message is received. The Follower remains in
-"connecting" until it receives `reconnect`, then it stays in "dropping" until
-it finishes halting all outstanding connections, after which it sends
-`reconnecting` and switches back to "connecting".
-
-"Connection hints" are type/address/port records that tell the other side of
-likely targets for L2 connections. Both sides will try to determine their
-external IP addresses, listen on a TCP port, and advertise `(tcp,
-external-IP, port)` as a connection hint. The Transit Relay is also used as a
-(lower-priority) hint. These are sent in `connection-hint` records, which can
-be sent any time after both sending and receiving a `please` record. Each
-side will initiate connections upon receipt of the hints.
+derived key ~warner — please please don't ~piegames)
 
-```
-{ "type": "connection-hints",
-  "hints": [ ... ]
-}
-```
+(TODO: reduce the number of round-trip stalls here, I've added too many ~warner)
+(yes, and one could also drastically reduce the number of messages by sending
+all hints at once with the reconnect message ~piegames)
 
-Hints can arrive at any time. One side might immediately send hints that can
-be computed quickly, then send additional hints later as they become
-available. For example, it might enumerate the local network interfaces and
-send hints for all of the LAN addresses first, then send port-forwarding
-(UPnP) requests to the local router. When the forwarding is established
-(providing an externally-visible IP address and port), it can send additional
-hints for that new endpoint. If the other peer happens to be on the same LAN,
-the local connection can be established without waiting for the router's
-response.
+Hints can arrive at any time. One side might immediately send hints that can be computed quickly,
+then send additional hints later as they become available.
+For example, it might enumerate the local network interfaces and send hints for all of the LAN addresses first,
+then send port-forwarding (UPnP) requests to the local router.
+When the forwarding is established (providing an externally-visible IP address and port),
+it can send additional hints for that new endpoint.
+If the other peer happens to be on the same LAN,
+the local connection can be established without waiting for the router's response.
 
+(I'd like to see that feature removed, this stuff is already complicated enough with a signle hints message ~piegames)
```

Comment:
To be clear, I meant "hints exchange". Of course there may be a lot of hints, but I was talking about each side sending only one message. Transit does it like that currently and it already is complicated enough.

Reply:
Not the state machine per se, but other things like for example prioritizing hints and merging duplicate hints. Currently you can implement it as one task like "have some hints now go and get me a connection", but with this you need to account for a way to side-load new hints into the process.

Reply:
I don't see any way to avoid that, and don't think it's desirable anyway as the state-machine will look approximately the same while making the application slower for users (and less flexible).